# CODIV-19

In [1]:
import numpy as np
import pandas as pd
from scipy.integrate import odeint
import torch
import pyro
import pyro.distributions as dist
from pyro.optim import Adam
from pyro.infer import SVI, Trace_ELBO
import pyro.poutine as poutine
from pyro.distributions import constraints

/home/lucy/Repositories/propios/ds-projects/neural-pdes-solver/.venv-np/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
# def sir_step(S, I, R, beta, gamma, N):
#     dS = -beta * S * I / N
#     dI = beta * S * I / N - gamma * I
#     dR = gamma * I
#     next_S = S + dS
#     next_I = I + dI
#     next_R = R + dR
#     return next_S, next_I, next_R

In [3]:
# Define the SIR differential equations
def sir_model(y, t, beta, gamma):
    S, I, R = y
    dSdt = -beta * S * I
    dIdt = beta * S * I - gamma * I
    dRdt = gamma * I
    return [dSdt, dIdt, dRdt]

In [4]:
# Integrate the SIR equations over the time grid using given parameters
def sir_solution(y0, t, beta, gamma):
    # Ensure tensors are detached from the current graph
    t_np = t.detach().numpy()
    y0_np = [y.detach().numpy() for y in y0]
    beta_np = beta.detach().numpy()
    gamma_np = gamma.detach().numpy()
    return odeint(sir_model, y0_np, t_np, args=(beta_np, gamma_np))

In [5]:
def model(susceptible, infected, recovered):
    # Priors sobre beta y gamma (unidades: dias^-1, no cambian al normalizar por poblacion)
    beta = pyro.sample("beta", dist.Uniform(0.05, 0.5))
    gamma = pyro.sample("gamma", dist.Uniform(0.05, 0.5))

    # Condiciones iniciales (ahora en fracciones de poblacion, N ~= 1)
    S0, I0, R0 = susceptible[0], infected[0], recovered[0]
    N = S0 + I0 + R0

    # Integracion Euler del SIR
    S, I, R = S0, I0, R0
    S_pred = [S0]
    I_pred = [I0]
    R_pred = [R0]

    for _ in range(1, len(susceptible)):
        dS = -beta * S * I / N
        dI = beta * S * I / N - gamma * I
        dR = gamma * I

        S = torch.clamp(S + dS, min=0.0)
        I = torch.clamp(I + dI, min=0.0)
        R = torch.clamp(R + dR, min=0.0)

        S_pred.append(S)
        I_pred.append(I)
        R_pred.append(R)

    S_pred = torch.stack(S_pred)
    I_pred = torch.stack(I_pred)
    R_pred = torch.stack(R_pred)

    # Ruido de observacion en FRACCION de poblacion (1e-3 ~ 328k personas).
    # El sigma=1000 original estaba en conteos absolutos frente a S~3.28e8,
    # lo que aplastaba la escala del likelihood.
    noise_scale = 1e-3
    with pyro.plate("data", len(susceptible)):
        pyro.sample("obs_S", dist.Normal(S_pred, noise_scale), obs=susceptible)
        pyro.sample("obs_I", dist.Normal(I_pred, noise_scale), obs=infected)
        pyro.sample("obs_R", dist.Normal(R_pred, noise_scale), obs=recovered)


In [6]:
# Load the data
us_daily_data = pd.read_csv("data/covid/us_covid19_daily.csv")

In [7]:
# Convert date columns to datetime format
us_daily_data['date'] = pd.to_datetime(us_daily_data['date'], format="%Y%m%d")

In [8]:
# Extract relevant columns for SIR modeling
sir_data = us_daily_data[['date', 'positive', 'death']].sort_values(by='date').reset_index(drop=True)

In [9]:
# Calculate the number of recovered individuals
recovery_delay = 14
sir_data['recovered'] = sir_data['positive'].shift(recovery_delay) - sir_data['death'].shift(recovery_delay)
sir_data['recovered'] = sir_data['recovered'].fillna(0)


In [10]:
# Fill with 0 the missing values of deaths
sir_data['death'] = sir_data['death'].fillna(0)


In [11]:
# For simplicity, we'll consider the entire US population as susceptible initially
us_population = 328_200_000  # US population estimate
sir_data['susceptible'] = us_population - sir_data['positive'] - sir_data['recovered']

# Recorte critico: eliminar filas anteriores al primer caso confirmado.
# Con I(0) = 0 la dinamica SIR queda congelada (dS = dI = dR = 0 para
# cualquier beta/gamma) y los gradientes de SVI son exactamente nulos:
# era la causa real de la perdida plana y de beta = gamma = 0.2.
first_case_idx = sir_data.index[sir_data['positive'] > 0][0]
sir_data = sir_data.loc[first_case_idx:].reset_index(drop=True)


In [12]:
# Convert data to PyTorch tensors, NORMALIZADOS como fraccion de la poblacion.
# Trabajar en conteos absolutos (S ~ 3.28e8) daba un likelihood de escala
# ~1e7 por punto; en fracciones (S ~ 1) los gradientes son utilizables.
susceptible_t = torch.tensor(sir_data['susceptible'].values, dtype=torch.float) / us_population
infected_t = torch.tensor(sir_data['positive'].values, dtype=torch.float) / us_population
recovered_t = torch.tensor(sir_data['recovered'].values, dtype=torch.float) / us_population


In [13]:
def guide(susceptible, infected, recovered):
    # Constraint interval(0.05, 0.5): sigue siendo coherente tras normalizar,
    # porque beta y gamma se miden en dias^-1 y no dependen de la escala de poblacion.
    beta_loc = pyro.param('beta_loc',
                         torch.tensor(0.2),
                         constraint=dist.constraints.interval(0.05, 0.5))
    gamma_loc = pyro.param('gamma_loc',
                          torch.tensor(0.2),
                          constraint=dist.constraints.interval(0.05, 0.5))

    # Sample parameters (guia Delta -> estimacion MAP)
    beta = pyro.sample('beta', dist.Delta(beta_loc))
    gamma = pyro.sample('gamma', dist.Delta(gamma_loc))


In [14]:
# Verificacion de la normalizacion: S+I+R ~= 1 y I(0) > 0 (dinamica viva)
print(f"n = {len(susceptible_t)} dias (desde el primer caso confirmado)")
print(f"S0 = {susceptible_t[0]:.6f}, I0 = {infected_t[0]:.3e}, R0 = {recovered_t[0]:.3e}")
print(f"I max = {infected_t.max():.6f} fraccion ({infected_t.max().item() * us_population:,.0f} personas)")


n = 183 dias (desde el primer caso confirmado)
S0 = 1.000000, I0 = 5.484e-08, R0 = 0.000e+00
I max = 0.018063 fraccion (5,928,381 personas)


In [15]:
# Setup for SVI
pyro.clear_param_store()
optimizer = Adam({"lr": 0.01})
svi = SVI(model, guide, optimizer, loss=Trace_ELBO())

In [16]:
# Running SVI
losses = []
num_iterations = 2000
for j in range(num_iterations):
    loss = svi.step(susceptible_t, infected_t, recovered_t)
    losses.append(loss)
    if j % 200 == 0:
        print("[iteration %04d] loss: %.4f" % (j + 1, loss / len(susceptible_t)))

print("[iteration %04d] loss: %.4f" % (num_iterations, losses[-1] / len(susceptible_t)))
losses[-1]


[iteration 0001] loss: 161.0701


[iteration 0201] loss: 63.7103


[iteration 0401] loss: 63.5796


[iteration 0601] loss: 63.3982


[iteration 0801] loss: 63.1749


[iteration 1001] loss: 62.9054


[iteration 1201] loss: 62.5980


[iteration 1401] loss: 62.2577


[iteration 1601] loss: 61.8902


[iteration 1801] loss: 61.5091


[iteration 2000] loss: 61.1316


11187.077667236328

In [17]:
# Extract the inferred parameters
beta_estimated = pyro.param("beta_loc").item()
gamma_estimated = pyro.param("gamma_loc").item()

print(f"Estimated beta: {beta_estimated}")
print(f"Estimated gamma: {gamma_estimated}")
print(f"R0 basico (beta/gamma): {beta_estimated / gamma_estimated:.3f}")

print()
print("FINAL_METRICS")
print(f"beta_estimado = {beta_estimated:.6f}")
print(f"gamma_estimado = {gamma_estimated:.6f}")
print(f"loss_inicial = {losses[0]:.4f}")
print(f"loss_final = {losses[-1]:.4f}")
print(f"loss_por_punto_inicial = {losses[0] / len(susceptible_t):.4f}")
print(f"loss_por_punto_final = {losses[-1] / len(susceptible_t):.4f}")
print(f"reduccion_loss_pct = {100 * (losses[-1] - losses[0]) / losses[0]:.1f}")
print(f"trayectoria_primeros5 = {[round(l, 2) for l in losses[:5]]}")
print(f"trayectoria_ultimos5 = {[round(l, 2) for l in losses[-5:]]}")
print(f"convergio = {losses[-1] < 0.9 * losses[0]}")


Estimated beta: 0.1930117905139923
Estimated gamma: 0.11941377818584442
R0 basico (beta/gamma): 1.616

FINAL_METRICS
beta_estimado = 0.193012
gamma_estimado = 0.119414
loss_inicial = 29475.8346
loss_final = 11187.0777
loss_por_punto_inicial = 161.0701
loss_por_punto_final = 61.1316
reduccion_loss_pct = -62.0
trayectoria_primeros5 = [29475.83, 29475.6, 29475.3, 29474.64, 29471.57]
trayectoria_ultimos5 = [11188.35, 11188.14, 11187.83, 11187.43, 11187.08]
convergio = True


## Resultado: el SVI converge tras corregir la condición inicial y la escala

**Actualización 2026-10-07 (segundo intento, ejecutado en Linux con pyro 1.9.2 / torch CPU).**
La inferencia **sí converge** (2000 iteraciones, Adam lr = 0.01): la pérdida ELBO cae de
**29475.83 → 11187.08** (−62.0 %; por punto: 161.07 → 61.13), de forma monótona, y los parámetros
se mueven claramente de su inicialización (0.2):

- **β estimado = 0.193012** días⁻¹
- **γ estimado = 0.119414** días⁻¹
- **R₀ básico (β/γ) ≈ 1.62**

Trayectoria de pérdida — primeros 5: [29475.83, 29475.60, 29475.30, 29474.64, 29471.57];
últimos 5: [11188.35, 11188.14, 11187.83, 11187.43, 11187.08].

### Qué hizo la diferencia

1. **Condición inicial no degenerada (causa principal del fallo original).** Los 38 primeros
   registros del dataset tienen `positive = 0`, lo que fijaba I(0) = 0. Con cero infectados la
   dinámica SIR queda congelada (dS = dI = dR = 0 para *cualquier* β y γ), así que los gradientes
   del ELBO eran exactamente nulos y los parámetros no podían moverse de 0.2 — la escala de σ era
   un problema secundario. Se recortaron las filas anteriores al primer caso confirmado
   (2020-02-29, 18 casos), quedando n = 183 días con I(0) = 5.5e-8 > 0.
2. **Normalización de las observaciones.** S, I, R ahora son fracciones de la población
   (÷ 328.2 M), y el ruido del likelihood pasó de σ = 1000 personas a σ = 1e-3 en fracción
   (≈ 328k personas). La pérdida por punto bajó de ~1.6e7 a O(10²), en un rango donde Adam
   recibe gradientes útiles.
3. **Constraint sin cambios.** Se mantiene `interval(0.05, 0.5)` en prior y guía: β y γ se miden
   en días⁻¹ y no dependen de la escala de población, así que ambos siguen siendo coherentes tras
   normalizar. La guía `Delta` (estimación MAP) también se mantiene.

*Nota histórica: la versión anterior de esta celda declaraba el fallo (pérdida plana ≈ 1.597e7 por
punto, β = γ = 0.2), verificado en la auditoría de 2026-10-07. El diagnóstico de entonces apuntaba
a la escala σ = 1000 frente a S ≈ 3.28e8; la re-ejecución muestra que la escala por sí sola no
bastaba — con I(0) = 0 la pérdida seguía plana (28794.82 constante) incluso ya normalizada.*
